refac: Migrate `CopyIcon` component from enzyme to RTL

EshaanAgg opened the pull request:
<!--
!! Please DELETE this comment before posting.
We appreciate your contribution to the Jaeger project! 👋🎉
-->

## Which problem is this PR solving?
Fixes part of #1668 

## Description of the changes
Migrates `CopyIcon` component from enzyme to RTL
- Introduced the `userEvent` library into the codebase with the standard for testing more complicated user interactions with RTL
- Removed the snapshot tests for `CopyIcon` component
- Decreased the test count of the component as the same when testing the `internal state of the component` which is against the policy of RTL. Instead replaced the same with a single comprehensive test which checks for the nature of the tooltip, and whether the content was copied into the clipboard or not

## How was this change tested?
Running the test suite locally 

## Checklist
- [X] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [X] I have signed all commits
- [X] I have added unit tests for the new functionality
- [X] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

yurishkuro commented on packages/jaeger-ui/src/components/common/CopyIcon.test.js:
```diff
@@ -28,43 +30,32 @@ describe('<CopyIcon />', () => {
+    // Simulate the click
+    await user.click(screen.getByRole('button'));
+    await user.hover(screen.getByRole('button'));
+    expect(screen.getByText('Copied')).toBeInTheDocument();
```
is this test going to he flaky? Once `user.click` happens, the Copied title only remains visible for a certain time - during CI the execution of this code might be suspended (CPU scheduling etc) and resumed after the timeout when "Copied" is no longer displayed, causing this assertion to fail. Some stateful test would be more reliable here, i.e. validating that a certain function was called (and separately testing that the tooltip changes based on state by _controlling the state_, not the user click behavior)

EshaanAgg replied:
- The test wouldn't be flaky as the tooltip would be visible for time-periods in magnitude of seconds, where as the suspension of the test shouldn't be more than microseconds. We are using `async` functions here as there is some time needed by React to render the elements into the DOM, which is not immediate. This is even true for the functionality in production code, and thus we needn't worry about it.

- I get the sentiment to feel more confident about the component using stateful tests, but the same goes against the philosophy of RTL itself. RTL is so heavily invested on testing user interactions, that it offers no utilities to get access to the `props` or `state` of the component after it's render. All you can do is use `querySelectors` to check for the appearance of elements. styles and classes.

yurishkuro replied:
1. It's possible for VM running CI to get suspended for seconds. However, on a second glance, the change in text does not happen on the timer, but on leaving the tooltip hover, which doesn't seem like it would happen in this test, so it's probably ok.
2. I think you can always construct code in a way to work around that. For example, the state change here happens via `handleTooltipVisibilityChange` function, that could be mocked if necessary.

Anyway, I think we're good because of (1) above.

yurishkuro replied:
Actually, you don't seem to test that the text will change back from Copied to Copy, is that something we can do by undoing hover state?

EshaanAgg replied:
> Actually, you don't seem to test that the text will change back from Copied to Copy, is that something we can do by undoing hover state?

Yes. I was thinking of adding that myself when I saw that I missed the `handleTooltipVisibilityChange` function in coverage. Will update that as well.

EshaanAgg commented on packages/jaeger-ui/src/components/common/CopyIcon.test.js:
```diff
@@ -28,43 +30,32 @@ describe('<CopyIcon />', () => {
+  it('updates state and copies when clicked', async () => {
+    // Intially there should be nothing in the clipboard, and thus the tooltip should have the default title
+    await user.hover(screen.getByRole('button'));
+    await waitFor(() => expect(screen.getByText('tooltipTitleValue')).toBeInTheDocument());
```
This is the title of the tooltip that we are passing in as a `prop` to the component. This would show when we hover over the `CopyIcon` button.

yurishkuro replied:
I see. Then we should define `const testTooltip = 'tooltipTitleValue'` and use that.

EshaanAgg replied:
Sure will do.

EshaanAgg commented:
@yurishkuro 
```ts 
handleTooltipVisibilityChange = (visible: boolean) => {
  if (!visible && this.state.hasCopied) {
    this.setState({
      hasCopied: false,
    });
  }
};
```
Can you please let me know what is the use of this function? I can't seem to figure it out.

EshaanAgg commented:
@yurishkuro I have added the `mouseLeave` test as you requested, but I can't seem to get it to pass. Can you look at the test and confirm that this is the intended behavior?

PS. I also removed the `userEvent` library for now, as the simple hover interactions can be done with `fireEvent` itself. We may include the library later when more complex interactions arise.

yurishkuro commented on packages/jaeger-ui/src/components/common/CopyIcon.test.js:
```diff
@@ -13,58 +13,57 @@
+
+    // Intially when we hover over the icon, there should be nothing in the clipboard and the tooltip should have the default title
+    fireEvent.mouseEnter(triggerElement);
+    await waitFor(() => expect(screen.getByText(tooltipTitle)).toBeInTheDocument());
```
what does waitFor do - is it periodically polling its lambda? Or does expect() always return a future? It's a bit strange that you suddenly start calling function in async which normally was not called this way.

Is `fireEvent.mouseEnter` sync or async?

EshaanAgg replied:
`waitFor` returns a Promise that resolves when the specified condition is met. It retries the condition every `50ms` (by default) until it passes or times out after `1000ms` (by default).

`fireEvent.mouseEnter` is a sync event, but even after the occurrence of the event, `antd` takes some time to insert the markdown for the tooltip into DOM. Thus we need to make use of the `waitFor` utility.
